### PR TITLE
Add --rm mode for inletsctl create

### LINKED ISSUE #41: Add a --rm flag to "inletsctl create" for a temporary tunnel
We should add a --rm flag to "inletsctl create" for when users need a temporary tunnel.

The use-case would be that I want to share my blog i.e. 127.0.0.1:4000 with a friend or colleague, but I don't want to be billed for the DigitalOcean VM beyond my 1-2 hours of uptime.

I'd run a command similar to `docker run --rm` which removes the tunnel immediately after I hit control + c or exit.

We'd have something like this, but it'd need to work with inlets-pro too:

OSS:

```
inletsctl create --rm --upstream http://127.0.0.1:4000

Your IP is: X
Starting "inlets client" now, hit control+c to delete the tunnel.
```

Pro:

```
inletsctl create --rm --license $LICENSE --remote-tcp=http://127.0.0.1
Your IP is: X
Starting "inlets-pro client" now, hit control+c to delete the tunnel.
```

### LINKED ISSUE #44: Fix issue with the opening of TCP ports for inlets-pro on GCE
## Expected Behaviour

The GCE operator should open up ports for inlets-pro
For users switching between inlets OSS and inlets-pro, the firewall rules should be 
updated automatically.

## Current Behaviour

If a firewall rule named inlets for inlets OSS already exists, it doesn't open up ports for inlets-pro

## Possible Solution

PR #45 

## Steps to Reproduce (for bugs)

1. Run inletsctl with inlets OSS
2. Run inletsctl again with inlets-pro

### PR DESCRIPTION
**Pull request by utsavanand2**
The --rm flag (default true) will enable to point to
an upstream or a remote-tcp and will delete the exit-node
on a SIGINT (control + c)

Fixes #41

Signed-off-by: Utsav Anand <utsavanand2@gmail.com>

## Description
The --rm flag will enable developers to quickly expose locally running endpoints
and delete the exit nodes when they are done previewing with a `ctrl + c`

## How are existing users impacted? What migration steps/scripts do we need?
Users will be able to use the --rm flag to delete the tunnel as soon as inletsctl is interrupted with
a (control + c). This will make sure any instances are deleted automatically after inletsctl exits with an interrupt.
A great feature for Web-Devs who want to give a preview of their projects with reviewers for a short while

## Checklist:

I have:

- [x] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/inlets/inlets/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

### CODE REVIEW

File: cmd/create.go [hunk 1]
```diff
@@ -180,46 +200,87 @@ func runCreate(cmd *cobra.Command, _ []string) error {
...
-  inlets client --remote "ws://%s:%d" \
-	--token "%s" \
-	--upstream $UPSTREAM
-
-To Delete:
-	inletsctl delete --provider %s --id "%s"
-`,
-					hostStatus.IP, inletsToken, hostStatus.IP, inletsControlPort, inletsToken, provider, hostStatus.ID)
-				return nil
+			if delTunnel == true {
```
Comment (alexellis) [inline cmd/create.go]: The functions look a little messy now doing now things than necessary. Could you refactor a little? How about a separate method to print the output of the command?

File: cmd/create.go [hunk 2]
```diff
@@ -180,46 +200,87 @@ func runCreate(cmd *cobra.Command, _ []string) error {
...
+				sig := make(chan os.Signal, 1)
+				done := make(chan bool, 1)
+
+				signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+
+				go func() {
```
Comment (alexellis) [inline cmd/create.go]: How does this compare to the implementation in the OpenFaaS watchdog? https://github.com/openfaas/faas/blob/master/watchdog/main.go

File: cmd/create.go (hunk 1 above)
Comment (utsavanand2) [reply cmd/create.go]: Yes that would be much better

File: cmd/create.go (hunk 2 above)
Comment (utsavanand2) [reply cmd/create.go]: I think I am not closing the channel here. It should be gracefully closed. As far as I can tell that is the only difference I can see

**Comment by alexellis** [comment]
```
./inletsctl create --rm --provider digitalocean --access-token  ~/Downloads/do-access-token
Using provider: digitalocean
Requesting host: recursing-leakey2 in , from digitalocean
2020/02/05 20:01:53 Provisioning host with DigitalOcean
```

I don't think this shouldn't work when the user hasn't specified an upstream port, can you show an error?

**Comment by alexellis** [comment]
Also can you think why I got an auth error?

```
POST https://api.digitalocean.com/v2/droplets: 401 Unable to authenticate you.
```

**Comment by alexellis** [comment]
Hmm it was `--access-token` vs `--access-token-file`

**Comment by alexellis** [comment]
```
  inletsctl create  \
	--provider [digitalocean|packet|ec2|scaleway|civo|gce] \
	--access-token-file $HOME/access-token \
	--region lon1
```

Please give an example showing --rm and how to invoke it with port in the CMD help text?

**Comment by alexellis** [comment]
The inlets-pro error is perhaps because the server takes a little while to come up, in which case you should retry the tunnel for let's say 5 times with a small pause between?

**Comment by utsavanand2** [comment]
Thanks a lot! I'll keep this feedback in mind for the next commit push.

**Comment by utsavanand2** [comment]
> ```
> ./inletsctl create --rm --provider digitalocean --access-token  ~/Downloads/do-access-token
> Using provider: digitalocean
> Requesting host: recursing-leakey2 in , from digitalocean
> 2020/02/05 20:01:53 Provisioning host with DigitalOcean
> ```
> 
> I don't think this shouldn't work when the user hasn't specified an upstream port, can you show an error?

I'm using a default value for the upstream, http://127.0.0.1:3000, I think we should log it while creating the tunnel saying, using the default value for upstream http://127.0.0.1:3000

What do you think?

**Comment by alexellis** [comment]
The output needs some work too, we should be using StreamStdio so that we get a live printout of the child process.

Instead I only see this when I hit control + c:

```
ct to proxy\" error=\"dial tcp 159.65.83.139:8080: connect: connection refused\"\ntime=\"2020-02-05T20:08:15Z\" level=error msg=\"Failed to connect to proxy\" error=\"dial tcp 159.65.83.139:8080: connect: connection refused\"\ntime=\"2020-02-05T20:08:20Z\" level=info msg=\"Connecting to proxy\" url=\"ws://159.65.83.139:8080/tunnel\"\n", exit-code: -1
Deleting tunnel
exiting
```

**Comment by alexellis** [comment]
No, don't use a default, if --rm is given, the user must change the port or specify remote-tcp.

**Comment by alexellis** [comment]
The behaviour looks different between create + delete, vs --rm then control+c - the ID wasn't printed out.

```
Deleting tunnel
exiting
kosmos:inletsctl alex$ ./inletsctl delete --provider digitalocean --access-token-file  ~/inlets-token --id 178924410
Using provider: digitalocean
Deleting host: 178924410 from digitalocean
```

**Comment by alexellis** [comment]
Your version of `github.com/alexellis/go-execute` is too old, please update it?

Also - you'll need to use Command for only the first part of the text, and Args[] for everything else. Please can you change that?

**Comment by utsavanand2** [comment]
Sure @alexellis I'll do this!

**Comment by utsavanand2** [comment]
Thanks a lot for the feedback!

**Comment by alexellis** [comment]
You're going to need a rebase now, because I've added go modules.

After rebasing from master, run:
```
go mod tidy
```

**Comment by alexellis** [comment]
Hi @utsavanand2 I noticed the issue title now says "fix issue with GCE running inlets-pro"

Didn't I merge that fix via another PR? If so, please could you rebase this PR against master and then edit the title to remove that part?

**Comment by utsavanand2** [comment]
@alexellis Sir I think I just need to change the title because this PR changes opens up all ports for running inlets-pro (specifically allowing all ports a user can use with `--tcp-ports` flag)

**Comment by alexellis** [comment]
I've edited the title for you on this PR to `(WIP) Add --rm mode for inletsctl create` -> it should not include any other changes. 

You also need a rebase. Could I ask you to prioritize fixing the following PRs #56 #44 #55? A separate branch might be a good idea.

Thank you for contributing :+1:

**Comment by utsavanand2** [comment]
Sure!

**Comment by utsavanand2** [comment]
@alexellis The latest commit solves an issue with getting to check if inlets-pro is running on a very small f1-micro instance on GCE before attempting to run the inlets client in pro mode.
Still it is not perfect for example if I do a `ctrl + c` while the status is being checked for the inlets-pro server to come online. The first message you get is `interrupt`, after pressing `ctrl + c` again you get a panic saying we sent a request to a closed channel

```
➜  inletsctl git:(tmp-tunnel) ✗ inletsctl create --rm -p gce -z us-central1-a --project-id uts-1382 -f ~/Desktop/uts-1382-ad5c53782d1d.json --remote-tcp=127.0.0.1 --tcp-ports=8080 -l=$LICENSE
Using provider: gce
Requesting host: jovial-saha8 in us-central1-a, from gce
2020/02/25 03:22:13 Creating firewall exists, updating: inlets
Host: jovial-saha8|us-central1-a|uts-1382, status: provisioning
Your IP is: 104.198.60.63
checking if the inlets server is up...
^C
interrupt
^Cpanic: send on closed channel

goroutine 4 [running]:
os/signal.process(0x226f980, 0xc000500650)
        /usr/local/opt/go/libexec/src/os/signal/signal.go:227 +0x189
os/signal.loop()
        /usr/local/opt/go/libexec/src/os/signal/signal_unix.go:23 +0x3e
created by os/signal.init.0
        /usr/local/opt/go/libexec/src/os/signal/signal_unix.go:29 +0x41
```

File: cmd/create.go [hunk 3]
```diff
@@ -200,49 +256,69 @@ func runCreate(cmd *cobra.Command, _ []string) error {
...
+					done <- true
+					if ok {
+						close(sig)
+					}
+				}()
+
+				fmt.Printf("Your IP is: %s\n", hostStatus.IP)
+
+				var err error = nil
+				ctx, cancelFunction := context.WithCancel(context.Background())
```
Comment (utsavanand2) [inline cmd/create.go]: @viveksyngh I would love to have your feedback on this. Even after passing a req context to checkServiceUp() we're not able to cancel our request and exit out.

**Comment by utsavanand2** [comment]
@viveksyngh I would love to have your feedback on the code, especially where we're creating and passing the context https://github.com/inlets/inletsctl/pull/45/files#r396123320

**Comment by alexellis** [comment]
@viveksyngh - I think there's been two members meetings where you've agreed to take a look at this, are you still able to assist or should we find someone else?

**Comment by viveksyngh** [comment]
I will have look at it this weekend.

**Comment by viveksyngh** [comment]
I have created this PR to utsav branch to fix the issue which @utsavanand2 is facing. 

cc @alexellis 

https://github.com/utsavanand2/inletsctl/pull/1

**Comment by utsavanand2** [comment]
@alexellis I think we're ready to merge this PR now so we can take advantage of the new feature

**Comment by utsavanand2** [comment]
@viveksyngh Thanks a lot for helping me out!

**Pull request closed without merging**